#  UR5 Manipulator Sensor Data Exploration – Notebook 01

**Objective:**  
This notebook performs **initial data exploration and structural validation** for UR5 robotic manipulator controller logs.  

**Data Sources:**
- `ur5testresult_header.xlsx`: Defines column names, types, and meanings.
- `calculated_deviation_of_actual_position_to_nominal_position.xlsx`: Contains deviation summaries (validation reference).
- Multiple `.csv` sensor data files: Controller-level logs at 125Hz across various payloads and test conditions.


##  Step 1: Import Dependencies and Define Paths
We start by loading essential libraries and defining directory paths for raw and processed data.


In [1]:
# --- Step 1: Imports & Paths ---
import pandas as pd
import numpy as np
import os
import glob

# Display settings
pd.set_option('display.max_columns', None)
pd.set_option('display.width', 120)

# Paths
header_path = "../data/raw/header/ur5testresult_header.xlsx"
sensor_path = "../data/raw/sensor_data/*.csv"
summary_path = "../data/raw/summary/calculated_deviation_of_actual_position_to_nominal_position.xlsx"

##  Step 2: Load and Inspect Header Metadata
This file contains descriptions and definitions for each column in the sensor dataset.  
It’s used to validate data consistency and understand what each signal represents.


In [2]:
# --- Step 2: Header Metadata ---
header_meta = pd.read_excel(header_path)
print(f"Header metadata shape: {header_meta.shape}")
header_meta.head()

Header metadata shape: (0, 73)


Empty DataFrame
Columns: [ROBOT_TIME, ROBOT_TARGET_JOINT_POSITIONS (J1), ROBOT_TARGET_JOINT_POSITIONS (J2), ROBOT_TARGET_JOINT_POSITIONS (J3), ROBOT_TARGET_JOINT_POSITIONS (J4), ROBOT_TARGET_JOINT_POSITIONS (J5), ROBOT_TARGET_JOINT_POSITIONS (J6), ROBOT_ACTUAL_JOINT_POSITIONS (J1), ROBOT_ACTUAL_JOINT_POSITIONS (J2), ROBOT_ACTUAL_JOINT_POSITIONS (J3), ROBOT_ACTUAL_JOINT_POSITIONS (J4), ROBOT_ACTUAL_JOINT_POSITIONS (J5), ROBOT_ACTUAL_JOINT_POSITIONS (J6), ROBOT_TARGET_JOINT_VELOCITIES (J1), ROBOT_TARGET_JOINT_VELOCITIES (J2), ROBOT_TARGET_JOINT_VELOCITIES (J3), ROBOT_TARGET_JOINT_VELOCITIES (J4), ROBOT_TARGET_JOINT_VELOCITIES (J5), ROBOT_TARGET_JOINT_VELOCITIES (J6), ROBOT_ACTUAL_JOINT_VELOCITIES (J1), ROBOT_ACTUAL_JOINT_VELOCITIES (J2), ROBOT_ACTUAL_JOINT_VELOCITIES (J3), ROBOT_ACTUAL_JOINT_VELOCITIES (J4), ROBOT_ACTUAL_JOINT_VELOCITIES (J5), ROBOT_ACTUAL_JOINT_VELOCITIES (J6), ROBOT_TARGET_JOITN_CURRENT (J1), ROBOT_TARGET_JOITN_CURRENT (J2), ROBOT_TARGET_JOITN_CURRENT (J3), ROBOT_TARGET_JOITN_CURRENT (J4), ROBOT_TARGET_JOITN_CURRENT (J5), ROBOT_TARGET_JOITN_CURRENT (J6), ROBOT_ACTUAL_JOINT_CURRENT (J1), ROBOT_ACTUAL_JOINT_CURRENT (J2), ROBOT_ACTUAL_JOINT_CURRENT (J3), ROBOT_ACTUAL_JOINT_CURRENT (J4), ROBOT_ACTUAL_JOINT_CURRENT (J5), ROBOT_ACTUAL_JOINT_CURRENT (J6), ROBOT_TARGET_JOINT_ACCELERATIONS (J1), ROBOT_TARGET_JOINT_ACCELERATIONS (J2), ROBOT_TARGET_JOINT_ACCELERATIONS (J3), ROBOT_TARGET_JOINT_ACCELERATIONS (J4), ROBOT_TARGET_JOINT_ACCELERATIONS (J5), ROBOT_TARGET_JOINT_ACCELERATIONS (J6), ROBOT_TARGET_JOINT_TORQUES (J1), ROBOT_TARGET_JOINT_TORQUES (J2), ROBOT_TARGET_JOINT_TORQUES (J3), ROBOT_TARGET_JOINT_TORQUES (J4), ROBOT_TARGET_JOINT_TORQUES (J5), ROBOT_TARGET_JOINT_TORQUES (J6), ROBOT_JOINT_CONTROL_CURRENT (J1), ROBOT_JOINT_CONTROL_CURRENT (J2), ROBOT_JOINT_CONTROL_CURRENT (J3), ROBOT_JOINT_CONTROL_CURRENT (J4), ROBOT_JOINT_CONTROL_CURRENT (J5), ROBOT_JOINT_CONTROL_CURRENT (J6), ROBOT_CARTESIAN_COORD_TOOL (x), ROBOT_CARTESIAN_COORD_TOOL (y), ROBOT_CARTESIAN_COORD_TOOL (z), ROBOT_CARTESIAN_COORD_TOOL (rx), ROBOT_CARTESIAN_COORD_TOOL (ry), ROBOT_CARTESIAN_COORD_TOOL (rz), ROBOT_TCP_FORCE (x), ROBOT_TCP_FORCE (y), ROBOT_TCP_FORCE (z), ROBOT_TCP_FORCE (rx), ROBOT_TCP_FORCE (ry), ROBOT_TCP_FORCE (rz), ROBOT_JOINT_TEMP (J1), ROBOT_JOINT_TEMP (J2), ROBOT_JOINT_TEMP (J3), ROBOT_JOINT_TEMP (J4), ROBOT_JOINT_TEMP (J5), ROBOT_JOINT_TEMP (J6)]
Index: []

In [3]:
print("\n--- Full List of Headers from the Excel File ---")
# Print all column names directly
print(header_meta.columns.tolist())


--- Full List of Headers from the Excel File ---
['ROBOT_TIME', 'ROBOT_TARGET_JOINT_POSITIONS (J1)', 'ROBOT_TARGET_JOINT_POSITIONS (J2)', 'ROBOT_TARGET_JOINT_POSITIONS (J3)', 'ROBOT_TARGET_JOINT_POSITIONS (J4)', 'ROBOT_TARGET_JOINT_POSITIONS (J5)', 'ROBOT_TARGET_JOINT_POSITIONS (J6)', 'ROBOT_ACTUAL_JOINT_POSITIONS (J1)', 'ROBOT_ACTUAL_JOINT_POSITIONS (J2)', 'ROBOT_ACTUAL_JOINT_POSITIONS (J3)', 'ROBOT_ACTUAL_JOINT_POSITIONS (J4)', 'ROBOT_ACTUAL_JOINT_POSITIONS (J5)', 'ROBOT_ACTUAL_JOINT_POSITIONS (J6)', 'ROBOT_TARGET_JOINT_VELOCITIES (J1)', 'ROBOT_TARGET_JOINT_VELOCITIES (J2)', 'ROBOT_TARGET_JOINT_VELOCITIES (J3)', 'ROBOT_TARGET_JOINT_VELOCITIES (J4)', 'ROBOT_TARGET_JOINT_VELOCITIES (J5)', 'ROBOT_TARGET_JOINT_VELOCITIES (J6)', 'ROBOT_ACTUAL_JOINT_VELOCITIES (J1)', 'ROBOT_ACTUAL_JOINT_VELOCITIES (J2)', 'ROBOT_ACTUAL_JOINT_VELOCITIES (J3)', 'ROBOT_ACTUAL_JOINT_VELOCITIES (J4)', 'ROBOT_ACTUAL_JOINT_VELOCITIES (J5)', 'ROBOT_ACTUAL_JOINT_VELOCITIES (J6)', 'ROBOT_TARGET_JOITN_CURRENT (J1)', 

##  Step 3: Detect and Inspect Raw Sensor Data Files
All CSV files are located in `../data/raw/sensor_data/`.  
Each file likely corresponds to a test run or a specific measurement scenario.
We’ll read and combine them into a single DataFrame for analysis.
Initial inspection helps understand data structure.

In [4]:
# --- Step 3: Sensor Data ---
sensor_files = sorted(glob.glob(sensor_path))

print("✔ Total sensor CSV files detected:", len(sensor_files))
print("\nList of Detected Sensor Files:")
for file in sensor_files:
    print(file)

✔ Total sensor CSV files detected: 18

List of Detected Sensor Files:
../data/raw/sensor_data/ur5testresult_coldstart_fullspeed_payload4.5lb_1.csv
../data/raw/sensor_data/ur5testresult_coldstart_fullspeed_payload4.5lb_2.csv
../data/raw/sensor_data/ur5testresult_coldstart_fullspeed_payload4.5lb_3.csv
../data/raw/sensor_data/ur5testresult_coldstart_halfspeed_payload4.5lb_1.csv
../data/raw/sensor_data/ur5testresult_coldstart_halfspeed_payload4.5lb_2.csv
../data/raw/sensor_data/ur5testresult_coldstart_halfspeed_payload4.5lb_3.csv
../data/raw/sensor_data/ur5testresult_fullspeed_payload1.6lb_1_1.csv
../data/raw/sensor_data/ur5testresult_fullspeed_payload1.6lb_2.csv
../data/raw/sensor_data/ur5testresult_fullspeed_payload1.6lb_3.csv
../data/raw/sensor_data/ur5testresult_fullspeed_payload4.5lb_1.csv
../data/raw/sensor_data/ur5testresult_fullspeed_payload4.5lb_2.csv
../data/raw/sensor_data/ur5testresult_fullspeed_payload4.5lb_3.csv
../data/raw/sensor_data/ur5testresult_halfspeed_payload1.6lb_1.c

In [5]:
# Read all sensor CSVs
sensor_dfs = []
for file in sensor_files:
    df = pd.read_csv(file)
    df["source_file"] = file.split("/")[-1]  # Keep track of file origin
    sensor_dfs.append(df)

# Combine all into one DataFrame
sensor_data = pd.concat(sensor_dfs, ignore_index=True)

print(f"Loaded {len(sensor_files)} sensor files")
print(f"Combined dataset shape: {sensor_data.shape}")
display(sensor_data.head())


Loaded 18 sensor files
Combined dataset shape: (153640, 806)


([747.2479999999999]   [-26.880068716264294   -79.91160892471794   57.09539150546534   -157.7717644023623  \
0            ([747.256]   [-26.880068716264294           -79.911609           57.095392          -157.771764   
1            ([747.264]   [-26.880068716264294           -79.911609           57.095392          -157.771764   
2  ([747.2719999999999]   [-26.880068716264294           -79.911609           57.095392          -157.771764   
3             ([747.28]   [-26.880068716264294           -79.911609           57.095392          -157.771764   
4            ([747.288]   [-26.880068716264294           -79.911609           57.095392          -157.771764   

    -105.0096132254245   -44.72477900700451]   [-26.876620441547427   -79.91090832767539   57.096775121424834  \
0          -105.009613   -44.72477900700451]   [-26.876620441547427           -79.910225            57.096092   
1          -105.009613   -44.72477900700451]    [-26.87937983797211           -79.909542            57.097485   
2          -105.009613   -44.72477900700451]    [-26.87937983797211           -79.910908            57.096092   
3          -105.009613   -44.72477900700451]    [-26.87730346046443           -79.909542            57.096775   
4          -105.009613   -44.72477900700451]    [-26.87869681905511           -79.911591            57.096092   

    -157.77315241285316   -105.0075641686735    -44.72546202592151]   [0.0   0.0   0.0.1   0.0.2   0.0.3   0.0]  \
0           -157.773835          -105.008930   -44.724095988087505]   [0.0   0.0     0.0     0.0     0.0   0.0]   
1           -157.772469          -105.007564    -44.72477900700451]   [0.0   0.0     0.0     0.0     0.0   0.0]   
2           -157.771103          -105.006171    -44.72270262949682]   [0.0   0.0     0.0     0.0     0.0   0.0]   
3           -157.773152          -105.006854   -44.724095988087505]   [0.0   0.0     0.0     0.0     0.0   0.0]   
4           -157.771786          -105.008247    -44.72614504483851]   [0.0   0.0     0.0     0.0     0.0   0.0]   

   [0.0.1   0.0.4   -0.0   0.0.5   0.0.6  0.0].1   [-2.2329763972194327e-18   -2.2138144127730115  \
0    [0.0     0.0   -0.0     0.0     0.0    0.0]   [-2.2329763972194327e-18             -2.213814   
1    [0.0     0.0   -0.0     0.0     0.0    0.0]   [-2.2329763972194327e-18             -2.213814   
2    [0.0     0.0   -0.0     0.0     0.0    0.0]   [-2.2329763972194327e-18             -2.213814   
3    [0.0     0.0   -0.0     0.0     0.0    0.0]   [-2.2329763972194327e-18             -2.213814   
4    [0.0     0.0   -0.0     0.0     0.0    0.0]   [-2.2329763972194327e-18             -2.213814   

    -1.5893478394401743   -0.1629906069639267   0.00045060505107219365  0.0].2    [0.2398737221956253  \
0             -1.589348             -0.162991                 0.000451    0.0]    [0.2398737221956253   
1             -1.589348             -0.162991                 0.000451    0.0]   [0.23539011180400848   
2             -1.589348             -0.162991                 0.000451    0.0]   [0.23314829170703888   
3             -1.589348             -0.162991                 0.000451    0.0]   [0.25780820846557617   
4             -1.589348             -0.162991                 0.000451    0.0]   [0.24884097278118134   

    -3.4344537258148193   -1.869670033454895   -0.3095833659172058   -0.20893064141273499   -0.10675287991762161]  \
0             -3.436696            -1.836043             -0.323309              -0.178430   -0.10522784292697906]   
1             -3.432212            -1.840526             -0.312633              -0.178430   -0.11132800579071045]   
2             -3.432212            -1.833801             -0.292808              -0.201305   -0.09760263562202454]   
3             -3.429970            -1.840526             -0.306533              -0.189105   -0.11132800579071045]   
4             -3.443421            -1.847252             -0.301958              -0.202830   -0.11895321309566498]   

   [0.0.2   0.0.7  

###  Structural Assessment: Why We Must Define 13 Columns
The reason for defining exactly 13 structural columns is mandatory due to the non-standard format of the raw sensor data and the requirements of the parsing method.

- 1. The Non-Standard Data Problem
     The sensor data files are not standard CSVs. They store all data for a single timestamp on one line, structured as a complex Python tuple:

    (Time, [List of 6 values], [List of 6 values], ..., [List of 6 values]).

If we were to use standard loading methods like pd.read_csv(), Pandas would incorrectly split the data based on all commas (including those inside the lists), resulting in hundreds of meaningless columns (as you previously saw with the 806 column count).

- 2. The Structural Mandate: 13 Top-Level Fields
     To correctly load the data, we must use the ast.literal_eval function. This function safely parses the complex string into a single Python list of its distinct, top-level elements.

When the raw text is parsed, it consistently yields exactly 13 primary fields per row:

---

Total Columns After Expansion

When lists are expanded into joint-indexed columns:

1 scalar column → 1 column

12 list fields × 6 joint values each → 72 columns

Total columns = 1 + 72 = 73

This matches exactly with the 73 header names provided in:

***`ur5testresult_header.xlsx`***

---

In [6]:
import ast

# Define 13 GENERIC column names. This is mandatory to structure the parsed data 
# that results from reading 13 fields per line.
COLUMNS_STRUCTURAL_GENERIC = [f"C{i+1}" for i in range(13)]

all_data = []

print("Starting robust structural parsing and combination...")

# --- Core Parsing Loop ---
# This loop iterates through the 'sensor_files' list generated in the previous step
for fpath in sensor_files:
    with open(fpath, "r") as f:
        for line in f:
            # Strip parentheses and prepare the line for safe evaluation
            line = line.strip().lstrip('(').rstrip(')')
            try:
                # Safely parse the line into a list of 13 elements
                parsed = ast.literal_eval(f"[{line}]")
                all_data.append(parsed)
            except:
                # Skip any malformed lines
                continue

# Create the final structural DataFrame
sensor_data_structural = pd.DataFrame(all_data, columns=COLUMNS_STRUCTURAL_GENERIC)

# --- Inspection of Combined Data ---
print(f"Loaded {len(sensor_files)} sensor files")
print(f"Combined structural dataset shape: {sensor_data_structural.shape}")
print(f"Total Rows: {sensor_data_structural.shape[0]}")
print(f"Total Columns: {sensor_data_structural.shape[1]}")

print("\n--- Structural Data Inspection (First 5 Rows) ---")
display(sensor_data_structural.head())

Starting robust structural parsing and combination...
Loaded 18 sensor files
Combined structural dataset shape: (153658, 13)
Total Rows: 153658
Total Columns: 13

--- Structural Data Inspection (First 5 Rows) ---


C1                                                 C2  \
0  [747.2479999999999]  [-26.880068716264294, -79.91160892471794, 57.0...   
1            [747.256]  [-26.880068716264294, -79.91160892471794, 57.0...   
2            [747.264]  [-26.880068716264294, -79.91160892471794, 57.0...   
3  [747.2719999999999]  [-26.880068716264294, -79.91160892471794, 57.0...   
4             [747.28]  [-26.880068716264294, -79.91160892471794, 57.0...   

                                                  C3                              C4                               C5  \
0  [-26.876620441547427, -79.91090832767539, 57.0...  [0.0, 0.0, 0.0, 0.0, 0.0, 0.0]  [0.0, 0.0, -0.0, 0.0, 0.0, 0.0]   
1  [-26.876620441547427, -79.9102253087584, 57.09...  [0.0, 0.0, 0.0, 0.0, 0.0, 0.0]  [0.0, 0.0, -0.0, 0.0, 0.0, 0.0]   
2  [-26.87937983797211, -79.9095422898414, 57.097...  [0.0, 0.0, 0.0, 0.0, 0.0, 0.0]  [0.0, 0.0, -0.0, 0.0, 0.0, 0.0]   
3  [-26.87937983797211, -79.91090832767539, 57.09...  [0.0, 0.0, 0.0, 0.0, 0.0, 0.0]  [0.0, 0.0, -0.0, 0.0, 0.0, 0.0]   
4  [-26.87730346046443, -79.9095422898414, 57.096...  [0.0, 0.0, 0.0, 0.0, 0.0, 0.0]  [0.0, 0.0, -0.0, 0.0, 0.0, 0.0]   

                                                  C6                                                 C7  \
0  [-2.2329763972194327e-18, -2.2138144127730115,...  [0.2398737221956253, -3.4344537258148193, -1.8...   
1  [-2.2329763972194327e-18, -2.2138144127730115,...  [0.2398737221956253, -3.4366955757141113, -1.8...   
2  [-2.2329763972194327e-18, -2.2138144127730115,...  [0.23539011180400848, -3.4322118759155273, -1....   
3  [-2.2329763972194327e-18, -2.2138144127730115,...  [0.23314829170703888, -3.4322118759155273, -1....   
4  [-2.2329763972194327e-18, -2.2138144127730115,...  [0.25780820846557617, -3.4299700260162354, -1....   

                               C8                                                 C9  \
0  [0.0, 0.0, 0.0, 0.0, 0.0, 0.0]  [-2.6149525002185862e-17, -25.67417061620775, ...   
1  [0.0, 0.0, 0.0, 0.0, 0.0, 0.0]  [-2.6149525002185862e-17, -25.67417061620775, ...   
2  [0.0, 0.0, 0.0, 0.0, 0.0, 0.0]  [-2.6149525002185862e-17, -25.67417061620775, ...   
3  [0.0, 0.0, 0.0, 0.0, 0.0, 0.0]  [-2.6149525002185862e-17, -25.67417061620775, ...   
4  [0.0, 0.0, 0.0, 0.0, 0.0, 0.0]  [-2.6149525002185862e-17, -25.67417061620775, ...   

                                                 C10                                                C11  \
0  [0.22866468131542206, -3.4344537258148193, -1....  [-0.6377188174812196, 0.2775360293865634, 0.75...   
1  [0.2376319169998169, -3.4344537258148193, -1.8...  [-0.6377179484974931, 0.27754250895226273, 0.7...   
2  [0.2376319169998169, -3.4344537258148193, -1.8...  [-0.6377234520452051, 0.2775758667322003, 0.75...   
3  [0.2376319169998169, -3.4344537258148193, -1.8...  [-0.6377102675303948, 0.27756214867412954, 0.7...   
4  [0.2645336389541626, -3.4344537258148193, -1.8...  [-0.6377319481118858, 0.27754833705585435, 0.7...   

                                                 C12                                                C13  
0  [-25.23140494271829, 17.43970690181103, 6.5165...  [25.209991455078125, 26.71473503112793, 26.712...  
1  [-27.052612657185563, 17.178791785221726, 8.71...  [25.209991455078125, 26.71473503112793, 26.712...  
2  [-26.723421390438553, 16.973602855200063, 8.11...  [25.209991455078125, 26.71473503112793, 26.712...  
3  [-26.701637393048827, 17.399379345542844, 7.90...  [25.209991455078125, 26.71473503112793, 26.712...  
4  [-26.205690541384136, 17.76316899344647, 7.908...  [25.209991455078125, 26.71473503112793, 26.714...

In [7]:
print("\n--- Data Types and Memory Usage ---")
sensor_data_structural.info()


--- Data Types and Memory Usage ---
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 153658 entries, 0 to 153657
Data columns (total 13 columns):
 #   Column  Non-Null Count   Dtype 
---  ------  --------------   ----- 
 0   C1      153658 non-null  object
 1   C2      153658 non-null  object
 2   C3      153658 non-null  object
 3   C4      153658 non-null  object
 4   C5      153658 non-null  object
 5   C6      153658 non-null  object
 6   C7      153658 non-null  object
 7   C8      153658 non-null  object
 8   C9      153658 non-null  object
 9   C10     153658 non-null  object
 10  C11     153658 non-null  object
 11  C12     153658 non-null  object
 12  C13     153658 non-null  object
dtypes: object(13)
memory usage: 15.2+ MB


###  Cleaning Is Required to Remove Noise and Invalid Characters

- The raw files contain unwanted characters:

    - ( ) tuple wrappers 

    - leading/trailing spaces

    - inconsistent delimiters

    - '[' and ']' list wrappers

    - embedded commas

- These must be handled before expansion, otherwise:

    - Parsing fails 

    - Numeric conversion fails 

    - Merging with header fails 

- Cleaning ensures:

    - All rows are valid Python literals

    - No malformed list structures

    - All values are numeric after expansion

##  Step4 : Summary File Inspection

The summary file `calculated_deviation_of_actual_position_to_nominal_position.xlsx`  
contains deviation statistics between **actual and target joint positions**.  
This serves as a **validation reference** for future analysis and anomaly detection.


In [8]:
# --- Step 4: Summary data ---
summary_df = pd.read_excel(summary_path)
print("Summary file loaded successfully.\n")
display(summary_df.head())

Summary file loaded successfully.



Calculated deviation of actual position to nominal position           Unnamed: 1
0                                          Position            delta Distance (mm)
1                                         position 1                      0.156205
2                                         position 2                      0.163095
3                                         position 3                      0.189737
4                                         position 4                       0.17398

In [9]:
print("Sensor Data Columns:", len(sensor_data.columns))
print("Summary Columns:", len(summary_df.columns))

Sensor Data Columns: 806
Summary Columns: 2


### Understanding the Summary File and How to Use It
`calculated_deviation_of_actual_position_to_nominal_position.xlsx`

This summary file contains pre-computed deviation statistics that measure how far the robot’s actual joint positions drifted from the nominal (target) positions during calibration or reference tests.

It serves two important roles:

1. Baseline Reference for System Accuracy

    These deviations represent the robot’s expected baseline performance.

    We can use them to:

    - Evaluate if future sensor data deviates abnormally

    - Compare real-time joint positions against known acceptable offsets

    - Validate whether loaded sensor logs reflect normal operation
  
2. Calibration & Threshold Setting for ML Tasks

    If we later build:

    - Anomaly detection

    - Predictive maintenance models

    - Outlier filtering

    - Drift detection systems

    …this file provides ground-truth thresholds.

## Step 5: Save Structural Data for Next Notebooks

At this stage, the dataset is **structurally aligned but uncleaned**.  
It will be used as input for the next notebook (`02_data_cleaning.ipynb`).

In [10]:
# --- Step 5: Save both CSV and Parquet ---

# Output directory
output_dir = "../data/processed"
os.makedirs(output_dir, exist_ok=True)

structured_csv = os.path.join(output_dir, "structured_sensor_data.csv")
structured_parquet = os.path.join(output_dir, "structured_sensor_data.parquet")

sensor_data.to_csv(structured_csv, index=False)
sensor_data.to_parquet(structured_parquet, index=False)

print(f"Structured dataset saved at:\n- {structured_csv}\n- {structured_parquet}")


Structured dataset saved at:
- ../data/processed/structured_sensor_data.csv
- ../data/processed/structured_sensor_data.parquet


## Step 6: Assessment & Conclusion: Data Readiness
   This section completes the 01_data_exploration notebook by assessing the structurally parsed data and preparing for the next phase.

   ---
   
   ###  Summary of Findings
   - Structural Integrity: All raw sensor files were successfully parsed and combined into a single, structurally sound DataFrame with 13 columns (rows: $\approx 153,658$).
   - Data Type Status: The data is not yet ML-ready. The majority of the columns are stored as object type (Python lists) and cannot be used directly for modeling.
   - Readiness: The data is ready for the Cleaning and Transformation Phase.
   
   ### Conclusion & Next Step
   The structural DataFrame is the clean input for the next step.
   - Goal Achieved: The fundamental data acquisition and structural validation are complete.
   - Next Phase: The raw list columns must be expanded into $\mathbf{72}$ individual numerical features, and all resulting $\mathbf{73}$ columns must be converted to $\mathbf{\text{float64}}$.
   - File Saved: The 13-column structural dataset has been saved to ../data/processed/sensor_data_structural.parquet.

   ---
   
   **`Next Notebook → 02_data_cleaning.ipynb`**